### PR TITLE
TINKERPOP-2143 JavaScript GLV: Support browsers

### PR DESCRIPTION
**Pull request by spmallette**
https://issues.apache.org/jira/browse/TINKERPOP-2143

The original PR for this is at #1070. The changes were simply brought here for ongoing rebasing until someone picks up the work and completes it. As it stands there are failing tests on a standard build.

### CODE REVIEW

**Comment by jorgebay** [comment]
I'll review it asap.

**Comment by jorgebay** [comment]
cc @fdominik

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js [hunk 1]
```diff
@@ -23,7 +23,8 @@
 'use strict';
 
 const EventEmitter = require('events');
-const WebSocket = require('ws');
+//const WebSocket = require('ws');
```
Comment (jorgebay) [inline gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js:26]: This import should still be included.

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js [hunk 2]
```diff
@@ -119,55 +120,56 @@ class Connection extends EventEmitter {
...
           result: null
-        };
-      }
+    };
+    }
 
-      const message = Buffer.from(this._header + JSON.stringify(this._getRequest(requestId, bytecode, op, args, processor)));
-      this._ws.send(message);
-    }));
+    //const message = Buffer.from(this._header + JSON.stringify(this._getRequest(requestId, bytecode, op, args, processor)));
```
Comment (jorgebay) [inline gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js:163]: leftover.

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js [hunk 3]
```diff
@@ -119,55 +120,56 @@ class Connection extends EventEmitter {
...
+     const message = this._header + JSON.stringify(this._getRequest(requestId, bytecode, op, args, processor));
+      var buf = new ArrayBuffer(message.length); // 2 bytes for each char
+      var bufView = new Uint8Array(buf);
+      for (var i=0, strLen=message.length; i < strLen; i++) {
```
Comment (jorgebay) [inline gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js:167]: Probably this needs a `TextEncoder` or something like that for fast access: https://developer.mozilla.org/en-US/docs/Web/API/TextEncoder

There are buffer libraries that use built-in `Buffer` in Node.js and polyfills for browsers, we should use one of those.

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js [hunk 4]
```diff
@@ -119,55 +120,56 @@ class Connection extends EventEmitter {
...
-      });
-    });
+      this._ws.onopen = (() => {
+      this.isOpen = true;
+    if (this._pingEnabled) {
+      this._pingHeartbeat();
+    }
+    resolve();
+  });
+  });
```
Comment (jorgebay) [inline gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js:149]: Indentation is wrong.

**Comment by jorupp** [comment]
Something else to think about here - some of the new edge-compute stuff (like Cloudflare workers) run a Browser-like JS environment as well, so supporting browsers should indirectly support environments like that.

At least I think so - I get an error `"ws does not work in the browser. Browser clients must use the native WebSocket object"` when using the gremlin JS SDK in a cloudflare worker, so I'd hope that working in a browser would get it working there too.

**Comment by spmallette** [comment]
closing this PR for now since it's now heavily conflicted and had remaining items to be addressed. if there is interested in reviving this work, please feel free to open a fresh PR.

**Pull request closed without merging**